### PR TITLE
Add Chinese ZUC stream cipher

### PR DESCRIPTION
**Pull request by InfoHunter**
Years ago, I opened a PR in OpenSSL project supporting the Chinese ZUC algorithm (https://github.com/openssl/openssl/pull/6886). This PR is a port of that one, on BabaSSL this time.

ZUC is a keystream generation algorithm standardized by Chinese SCA as GM/T 0001.1-2012. ZUC is also standardized by 3GPP: [3GPP-ZUC](https://www.gsma.com/aboutus/wp-content/uploads/2014/12/eea3eia3zucv16.pdf)

128-EEA3 is standardized by 3GPP in its "TS 35.221" specification ([128-EEA3](https://www.gsma.com/aboutus/wp-content/uploads/2014/12/EEA3_EIA3_specification_v1_7.pdf)) as a stream cipher for using in 3G and LTE.

128-EEA3 is also known as "The ZUC-based confidentiality algorithm", which is standardized by SCA as GM/T 0001.2-2012.

The name ZUC is the abbreviation of an ancient Chinese mathematician "[Zu Chongzhi](https://en.wikipedia.org/wiki/Zu_Chongzhi)", so this algorithm is also called "Zu Chongzhi Algorithm" in China.

There is also an integrity algorithm based on ZUC, which is standardized by 3GPP as 128-EIA3. That one is not included in this PR.

### CODE REVIEW

**Comment by InfoHunter** [comment]
Note: I left the copyright line `Copyright 2018 BaishanCloud. All Rights Reserved` in the source file to honor BaishanCloud for supporting the development of this patch during year 2018...

**Comment by InfoHunter** [comment]
But I am considering if we should remove the OpenSSL copyright stuffs since this patch gets not merged by OpenSSL yet...and adding Alipay's copyright should be more proper maybe...

**Comment by InfoHunter** [comment]
WIP removed, ready for review now

File: include/openssl/sslerr.h [hunk 1]
```diff
@@ -830,6 +830,8 @@ int ERR_load_SSL_strings(void);
 # define SSL_R_INVALID_SRP_USERNAME                       357
 # define SSL_R_INVALID_STATUS_RESPONSE                    328
 # define SSL_R_INVALID_TICKET_KEYS_LENGTH                 325
+# define SSL_R_KEYLESS_ERROR                              295
```
Comment (dongbeiouba) [inline include/openssl/sslerr.h:833]: Please delete useless code.

File: ssl/ssl_err.c [hunk 2]
```diff
@@ -1416,6 +1416,9 @@ static const ERR_STRING_DATA SSL_str_reasons[] = {
     "invalid status response"},
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_INVALID_TICKET_KEYS_LENGTH),
     "invalid ticket keys length"},
+    {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_KEYLESS_ERROR), "keyless error"},
```
Comment (dongbeiouba) [inline ssl/ssl_err.c:1419]: Please delete useless code.

File: util/libssl.num [hunk 3]
```diff
@@ -564,12 +564,12 @@ SSL_use_sign_PrivateKey                 614	1_1_1e	EXIST::FUNCTION:NTLS,SM2
 SSL_use_enc_certificate                 615	1_1_1e	EXIST::FUNCTION:NTLS,SM2
 SSL_use_enc_certificate_file            616	1_1_1e	EXIST::FUNCTION:NTLS,SM2
 SSL_use_enc_PrivateKey                  617	1_1_1e	EXIST::FUNCTION:NTLS,SM2
-SSL_CTX_get_options2                    620	1_1_1g	EXIST::FUNCTION:OPTIONS2
```
Comment (dongbeiouba) [inline util/libssl.num:567]: Please delete useless code.

Comment (InfoHunter) [reply util/libssl.num:567]: Don't quite understand this...

**Comment by jinjiu** [comment]
LGTM

File: test/zuc_internal_test.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+        /* L */
+        2000,
+
+        /* output */
+        { 0xed4400e7, 0x0633e5c5, 0x7a574cdb }
+    }
+};
+
+static int test_zuc(int idx)
```
Comment (dongbeiouba) [inline test/zuc_internal_test.c:115]: What's the difference between these test cases? Can you add some comments?

File: util/libssl.num (hunk 3 above)
Comment (dongbeiouba) [reply util/libssl.num:567]: No options2 API, so no NOEXIST::FUNCTION:

File: crypto/evp/e_eea3.c [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+static int eea3_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
+                        const unsigned char *iv, int enc);
+static int eea3_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                      const unsigned char *in, size_t inl);
+static int eea3_cleanup(EVP_CIPHER_CTX *ctx);
+
+static const EVP_CIPHER zuc_128_eea3_cipher = {
+    NID_zuc_128_eea3,
+    1, EVP_ZUC_KEY_SIZE, 5,
```
Comment (dongbeiouba) [inline crypto/evp/e_eea3.c:30]: The length of iv is 5 bytes? should it be 16 bytes?

**Comment by InfoHunter** [comment]
One CI failure is caused by the doc nits checking script. Fixed in https://github.com/BabaSSL/BabaSSL/pull/30

**Comment by InfoHunter** [comment]
CIs are all happy now and all other comments are addressed as well.

File: crypto/evp/e_eea3.c (hunk 5 above)
Comment (InfoHunter) [reply crypto/evp/e_eea3.c:30]: EEA3 calls ZUC to generate the keystream, which requires a 16 bytes IV value. But the IV is not fully random, the IV is formed by three parameters: count, bearer and direction. OpenSSL's EVP framework doesn't support passing those parameters into a cipher, so I use the IV to hold those three parameters and then build the real IV inside EEA3...and then call ZUC to generate the keystream. The parameter count, bearer and direction require 38 bits to store, so 5 bytes (40 bits) are allocated here to hold those values.

Comment (InfoHunter) [reply crypto/evp/e_eea3.c:30]: Check `eea3_init_key()` function for details.

**Comment by dongbeiouba** [comment]
LGTM

**Pull request merged**